scanner/services/converters/converters.go

```diff
@@ -0,0 +1,400 @@
+	if err != nil {
+		return nil, fmt.Errorf("internal error: %w", err)
+	}
+	return &v4.VulnerabilityReport{
+		Vulnerabilities:        vulnerabilities,
+		PackageVulnerabilities: toProtoV4PackageVulnerabilitiesMap(r.PackageVulnerabilities),
+		Contents:               toProtoV4Contents(r.Packages, r.Distributions, r.Repositories, r.Environments),
+	}, nil
+}
+
+// ToClairCoreIndexReport converts v4.Contents to a claircore.IndexReport.
+func ToClairCoreIndexReport(contents *v4.Contents) (*claircore.IndexReport, error) {
+	if contents == nil {
+		return nil, errors.New("internal error: empty contents")
+	}
+	pkgs, err := convertSliceToMap(contents.GetPackages(), toClairCorePackage)
+	if err != nil {
+		return nil, fmt.Errorf("internal error: %w", err)
+	}
+	dists, err := convertSliceToMap(contents.GetDistributions(), toClairCoreDistribution)
+	if err != nil {
+		return nil, fmt.Errorf("internal error: %w", err)
+	}
+	repos, err := convertSliceToMap(contents.GetRepositories(), toClairCoreRepository)
+	if err != nil {
+		return nil, fmt.Errorf("internal error: %w", err)
+	}
+	var environments map[string][]*claircore.Environment
+	if envs := contents.GetEnvironments(); envs != nil {
+		environments = make(map[string][]*claircore.Environment, len(envs))
+		for k, v := range envs {
+			for _, env := range v.GetEnvironments() {
+				ccEnv, err := toClairCoreEnvironment(env)
+				if err != nil {
+					return nil, err
+				}
+				environments[k] = append(environments[k], ccEnv)
+			}
+		}
+	}
+	return &claircore.IndexReport{
+		Packages:      pkgs,
+		Distributions: dists,
+		Repositories:  repos,
+		Environments:  environments,
+	}, nil
+}
+
+func toProtoV4Contents(
+	pkgs map[string]*claircore.Package,
+	dists map[string]*claircore.Distribution,
+	repos map[string]*claircore.Repository,
+	envs map[string][]*claircore.Environment,
+) *v4.Contents {
+	var environments map[string]*v4.Environment_List
+	if len(envs) > 0 {
+		environments = make(map[string]*v4.Environment_List, len(envs))
+	}
+	for k, v := range envs {
+		l, ok := environments[k]
+		if !ok {
+			l = &v4.Environment_List{}
+			environments[k] = l
+		}
+		for _, e := range v {
+			l.Environments = append(l.Environments, toProtoV4Environment(e))
+		}
+	}
+	return &v4.Contents{
+		Packages:      convertMapToSlice(toProtoV4Package, pkgs),
+		Distributions: convertMapToSlice(toProtoV4Distribution, dists),
+		Repositories:  convertMapToSlice(toProtoV4Repository, repos),
+		Environments:  environments,
+	}
+}
+
+func toProtoV4Package(p *claircore.Package) *v4.Package {
+	if p == nil {
+		return nil
+	}
+	// Conversion functions.
+	toNormalizedVersion := func(version claircore.Version) *v4.NormalizedVersion {
+		return &v4.NormalizedVersion{
+			Kind: version.Kind,
+			V:    version.V[:],
+		}
+	}
+	toSourcePackage := func(src *claircore.Package) *v4.Package {
+		if src == nil {
+			return nil
+		}
+		// Sanitize and avoid recursion.
+		src.Source = nil
```

Comment:
This seems risky / may create some tough to troubleshoot bugs in the future, perhaps its worth adding a disclaimer to the function's docs stating the input will be modified.  Or perhaps it's worth creating a function that performs the conversion but ignores the source field?

Reply:
I agree this is cutting a corner, but can you clarify a scenario where this sanitation might cause problems? Asking because by definition, source packages should not have sources themselves, unless we are talking about some broader interpretation of source packages for non-OS language packages?

Reply:
I wasn't commenting on data specific scenario (I do not know the data well enough).  The scenario may not exist now for all I know, but if it happens to be valid in the future - this could potentially lead to a difficult bug.

Conventionally I wouldn't expect `clairReport` to be modified in any way if this was executed:
```go
acsVulnReport := converters.ToProtoV4VulnerabilityReport(clairReport)
```

Reply:
ClairCore's response to `Package.Source.Source`:

> we shouldn't ever have a source-source, correct

So sanitizing input here seems reasonable unless you want to block `Package.Source.Source != nil` as invalid input. Can you clarify?

Reply:
OK, thanks for clarifying - if we're confident this statement will hold true for the 'foreseeable future', then am OK with what you have.  If your extremely confident then could remove the `nil` assignment all together 😉 , but I'd rather see the defensive approach you have here vs. a crash.

My opinion (which will NOT hold up my approval here) is that we shouldn't ever modify the input object in a converter such as this - either populate the source without using recursion (preferred) or reject the input.

Reply:
Maybe the method should be named "Sanitizers" instead. My goal is to **validate and clean** input that I believe is worth cleaning so subsequent logic can work smoothly and abort (with an error) if something is not as expected.

Would you like to change the package and method names to "Sanitize"?

Reply:
Or do you prefer we should block the request altogether?

Reply:
Yes would rather see that (I know it's a trivial nuance) or clarify in the package docs that input may be modified.

Reply:
eb18c809db..7d8ddb9c0c